pkg/controller/rollouthistory/rollouthistory_controller.go

```diff
@@ -0,0 +1,393 @@
+	trafficRoutingInfo := rolloutv1alpha1.TrafficRoutingInfo{}
+	var err error
+	// if gateway is configured, get it
+	if rollout.Spec.Strategy.Canary.TrafficRoutings[0].Gateway != nil &&
+		rollout.Spec.Strategy.Canary.TrafficRoutings[0].Gateway.HTTPRouteName != nil {
+		trafficRoutingInfo.HTTPRoute, err = r.getGateWayInfo(rollout)
+		if err != nil {
+			return trafficRoutingInfo, err
+		}
+	}
+	// if ingress is configured, get it
+	if rollout.Spec.Strategy.Canary.TrafficRoutings[0].Ingress != nil &&
+		rollout.Spec.Strategy.Canary.TrafficRoutings[0].Ingress.Name != "" {
+		trafficRoutingInfo.Ingress, err = r.getIngressInfo(rollout)
+		if err != nil {
+			return trafficRoutingInfo, err
+		}
+	}
+	return trafficRoutingInfo, nil
+}
+
+// getGateWayInfo get HTTPRouteInfo for TrafficRoutingInfo
+func (r *RolloutHistoryReconciler) getGateWayInfo(rollout *rolloutv1alpha1.Rollout) (*rolloutv1alpha1.HTTPRouteInfo, error) {
+	// get HTTPRoute
+	gatewayName := *rollout.Spec.Strategy.Canary.TrafficRoutings[0].Gateway.HTTPRouteName
+	HTTPRoute := &v1alpha2.HTTPRoute{}
+	err := r.Get(context.TODO(), types.NamespacedName{Namespace: rollout.Namespace, Name: gatewayName}, HTTPRoute)
+	if err != nil {
+		return nil, errors.New("initGateway error: HTTPRoute " + gatewayName + " not find")
+	}
+	// marshal HTTPRoute into HTTPRouteInfo for rolloutHistory
+	gatewayInfo := &rolloutv1alpha1.HTTPRouteInfo{}
+	gatewayInfo.Name = HTTPRoute.Name
+	if gatewayInfo.Data.Raw, err = json.Marshal(HTTPRoute.Spec); err != nil {
+		return nil, err
+	}
+	return gatewayInfo, nil
+}
+
+// getIngressInfo get IngressInfo for TrafficRoutingInfo
+func (r *RolloutHistoryReconciler) getIngressInfo(rollout *rolloutv1alpha1.Rollout) (*rolloutv1alpha1.IngressInfo, error) {
+	// get Ingress
+	ingressName := rollout.Spec.Strategy.Canary.TrafficRoutings[0].Ingress.Name
+	ingress := &networkingv1.Ingress{}
+	err := r.Get(context.TODO(), types.NamespacedName{Namespace: rollout.Namespace, Name: ingressName}, ingress)
+	if err != nil {
+		return nil, errors.New("initIngressInfo error: Ingress " + ingressName + " not find")
+	}
+	// marshal ingress into ingressInfo
+	ingressInfo := &rolloutv1alpha1.IngressInfo{}
+	ingressInfo.Name = ingressName
+	if ingressInfo.Data.Raw, err = json.Marshal(ingress.Spec); err != nil {
+		return nil, err
+	}
+	return ingressInfo, nil
+}
+
+// updateRolloutHistoryWhenRolloutIsCompeleted record all pods released when rollout phase is healthy
+func (r *RolloutHistoryReconciler) updateRolloutHistoryWhenRolloutIsCompeleted(rollout *rolloutv1alpha1.Rollout, rolloutHistory *rolloutv1alpha1.RolloutHistory) error {
+	// do update until rollout.status.Phase is equl to RolloutPhaseHealthy
+	if rollout.Status.Phase != rolloutv1alpha1.RolloutPhaseHealthy {
+		return nil
+	}
+	// when this rollot's phase has been healthy, rolloutHistory record steps information and rollout.spec
+	// record .spec for rolloutHistory
+	var err error
```

Comment:
```
spec, err := r.getRolloutHistorySpec(rollout)
	if err != nil {
		return err
	}
	if !reflect.DeepEqual(rolloutHistory.Spec, spec) {
		rolloutHistory.Spec = spec
		err = r.Update(context.TODO(), rolloutHistory, &client.UpdateOptions{})
		if err != nil {
			return err
		}
		klog.Infof("")
		return nil
	}
```

Reply:
Thank you very much! Now I learn that a rolloutHistory will record .spec and .status, and it need requeue after update .spec. 
And I have udated unite_test just now. Before I just depends on the rollout event, for it will changed more than one time.